File: converter/zon.py
from buffer import Buffer
from ter import readTer
from mod import readMod
from zonefile import *

def readZon(data, zone, s3d):
    def getString(pos):
        return strs[pos:].split('\0', 1)[0]

    b = Buffer(data)
    assert b.read(4) == 'EQGZ'

    version, strlen = b.uint(2)
    numfiles, numplaceable, numunk3 = b.uint(3)
    numlights = b.uint()

    strs = b.read(strlen)

    files = [getString(b.uint()).lower() for i in xrange(numfiles)]
    objects = []
    for fn in files:
        if fn.endswith('.ter'):
            readTer(s3d[fn], zone, s3d)
            objects.append(None) # Shouldn't be placing zones...
        elif fn.endswith('.mod'):
            print 'reading mod', fn
            obj = zone.addObject(name=fn)
            #readMod(s3d[fn], obj, s3d)
            objects.append(obj)
    
    for i in xrange(numplaceable):
        #objname = files[b.uint()]
        objname = ''
        name = getString(b.uint())
        pos = b.float(3)
        ra, rb, rc = b.float(3)
        rot = (rc, rb, ra)
        scale = b.float()

        #zone.addPlaceable(objname, pos, rot, (scale, scale, scale))
    
    for i in xrange(numunk3):
        u1 = b.uint()
        u2 = b.float(9)
        #print getString(u1), u2
    
    #b.pos = len(b) - (32 * numplaceable)
    for i in xrange(numlights):
        obj = strs[b.uint():].split('\0', 1)[0]
        pos = b.float(3)
        rot = b.float(3)
        scale = b.float()
        #print obj, pos, rot, scale

File: LegacyFileReader/Zon.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Numerics;
using System.Text;
using MoreLinq;
using OpenEQ.Common;
using static System.Console;

namespace OpenEQ.LegacyFileReader {
	public class Zon {
		public readonly S3D S3D;

		public readonly List<TerMod> Objects;
		public readonly List<(int ObjId, string Name, Vector3 Position, Vector3 Rotation, float Scale)> Placeables;
		public readonly List<(string Name, Vector3 Position, Vector3 Color, float Radius)> Lights;
		
		public Zon(S3D s3d, Stream fp) {
			S3D = s3d;
			var br = new BinaryReader(fp, Encoding.Default, leaveOpen: true);

			var magic = br.ReadUInt32();
			Debug.Assert(magic == 0x5a475145); // 'EQGZ'
			var version = br.ReadUInt32();
			Debug.Assert(version == 1);
			var strlen = br.ReadInt32();
			var numFiles = br.ReadInt32();
			var numPlaceable = br.ReadInt32();
			var numUnk = br.ReadInt32();
			var numLights = br.ReadInt32();

			var strTable = Encoding.ASCII.GetString(br.ReadBytes(strlen));
			string GetString(int offset) => strTable.Substring(offset, strTable.IndexOf('\0', offset) - offset);

			Objects = Enumerable.Range(0, numFiles).Select(x => {
				var fn = GetString(br.ReadInt32()).ToLower();
				if(fn.EndsWith(".ter"))
					return new TerMod(s3d.Open(fn), true);
				if(fn.EndsWith(".mod"))
					return new TerMod(s3d.Open(fn), false);
				throw new NotImplementedException($"Unknown file reference in .zon: {fn}");
			}).ToList();

			Placeables = Enumerable.Range(0, numPlaceable).Select(x => {
				var objId = br.ReadInt32();
				var name = GetString(br.ReadInt32());
				var pos = br.ReadVec3();
				var rot = br.ReadVec3();
				var scale = br.ReadSingle();
				return (objId, name, pos, rot, scale);
			}).ToList();

			Enumerable.Range(0, numUnk).ForEach(x => {
				br.ReadUInt32();
				br.ReadVec3();
				br.ReadVec3();
				br.ReadVec3();
			});

			var lt = new Vector3(1, -1, 1);
			Lights = Enumerable.Range(0, numLights)
				.Select(x => (GetString(br.ReadInt32()), br.ReadVec3().YXZ() * lt, br.ReadVec3(), br.ReadSingle())).ToList();
		}
	}
}